src/fmc.rs

```diff
@@ -0,0 +1,564 @@
+
+pub trait SDRAMClockAndChipEnable {
+    fn sdram_bank_selection() -> SDRAMInternalBankSelection;
+    fn set_fmc_speed_and_pullup(self) -> Self;
+}
+
+pub struct UseSDRAMBank1<SDCKE0: PinSDCKE0, SDNE0: PinSDNE0>(pub SDCKE0, pub SDNE0);
+pub struct UseSDRAMBank2<SDCKE1: PinSDCKE1, SDNE1: PinSDNE1>(pub SDCKE1, pub SDNE1);
+
+impl<SDCKE0: PinSDCKE0, SDNE0: PinSDNE0> SDRAMClockAndChipEnable for UseSDRAMBank1<SDCKE0, SDNE0> {
+    fn sdram_bank_selection() -> SDRAMInternalBankSelection {
+        SDRAMInternalBankSelection::Bank1
+    }
+
+    fn set_fmc_speed_and_pullup(self) -> Self {
+        UseSDRAMBank1(
+            self.0.set_fmc_speed_and_pullup(),
+            self.1.set_fmc_speed_and_pullup(),
+        )
+    }
+}
+
+impl<SDCKE1: PinSDCKE1, SDNE1: PinSDNE1> SDRAMClockAndChipEnable for UseSDRAMBank2<SDCKE1, SDNE1> {
+    fn sdram_bank_selection() -> SDRAMInternalBankSelection {
+        SDRAMInternalBankSelection::Bank2
+    }
+
+    fn set_fmc_speed_and_pullup(self) -> Self {
+        UseSDRAMBank2(
+            self.0.set_fmc_speed_and_pullup(),
+            self.1.set_fmc_speed_and_pullup(),
+        )
+    }
+}
+
+pub struct SDRAMConfig {
+    /// width of the address bus
+    pub address_width: SDRAMAddressWidth,
+
+    /// width of the data bus
+    pub data_width: SDRAMDataWidth,
+
+    pub number_of_column_bits: NC_A,
+    pub number_of_row_bits: NR_A,
+
+    /// number of banks in module (BA0/BA1 pins)
+    pub module_bank_count: NB_A,
+
+    pub sdclk_period: SDCLK_A,
+    pub rpipe_delay: RPIPE_A,
+    pub write_protection: bool,
+    pub read_burst: bool,
+    pub timing: SDRAMTiming,
+}
+
+/// Timing configurations for SDRAM. All units in SDRAM clocks (confirm?)
+/// todo: use a nibble type for these if possible...
+pub struct SDRAMTiming {
+    pub cas_latency: CAS_A,
+    /// LOAD MODE REGISTER command to ACTIVE command delay (sometimes called t_mrd).
+    /// "Load mode register to active delay" in CubeMX
+    pub t_mrd: u8,
+    /// Exit SELF_REFRESH-to-ACTIVE command delay
+    /// "Exit self-refresh delay" in CubeMX
+    pub t_xsr: u8,
+    /// Self refresh time (as named in CubeMX, tRAS in some datasheets)
+    pub t_ras: u8,
+    /// ACTIVE-to-ACTIVE command period (sometimes called tRC)
+    /// "SDRAM common row cycle delay" in CubeMX
+    pub t_rc: u8,
+    /// WRITE recovery time (as named in CubeMX, tWR in datasheets)
+    pub t_wr: u8,
+    /// PRECHARGE command period (tRP in datasheets)
+    /// "SDRAM common row precharge delay" in CubeMX
+    pub t_rp: u8,
+    /// ACTIVE-to-READ or Write delay (tRCD in datasheets)
+    /// "Row to column delay" in CubeMX
+    pub t_rcd: u8,
+}
+
+impl SDRAMConfig {
+    fn configure_gpio<CCE: SDRAMClockAndChipEnable, SDNWE: PinSDNWE>(
+        &self,
+        cce: CCE,
+        sdnwe: SDNWE,
+    ) {
+        let peripherals = unsafe { crate::stm32::Peripherals::steal() };
+        let gpiod = peripherals.GPIOD.split();
```

Comment:
Those operations are risky as they do RMW operations. At the very least we need to somehow protect these. Also I think those GPIOs don't belong into something called `SDRAMConfig` as the GPIOs are completely independent of what exactly you're attaching to your parallel bus.

Probably better to move the GPIO configuration out into it's own thing. Also an interesting approach to explore would be to have the traditional `new` function where you have to pass in all the correct pins and a (unsafe) `new_unchecked` function which assumes that the application has properly set up everything.

Reply:
>Those operations are risky as they do RMW operations. At the very least we need to somehow protect these. 

Agree with the risk, I was hoping it's mitigated by the fact that the operations they do are idempotent (i.e., they never disable any GPIO peripherals or pins, only additively enable them)

>Also I think those GPIOs don't belong into something called SDRAMConfig as the GPIOs are completely independent of what exactly you're attaching to your parallel bus. Probably better to move the GPIO configuration out into it's own thing.

Do you mean the GPIOs which are passed into `configure`, or generally the notion of `impl SDRAMConfig` being ultimately responsible for setting up the GPIOs. If the former -- those GPIOs are necessary for SDRAM functionality. The are passed into `SDRAMConfig.configure(...)` which consumes the SDRAMConfig and returns an SDRAM. This is deliberate done so that at least ownership of the swappable SDRAM control pins is taken in this case. If the latter, it's only done that way to separate out the logic of the different steps (i.e., so we don't have a gigantic `configure()` function)

>Also an interesting approach to explore would be to have the traditional new function where you have to pass in all the correct pins and a (unsafe) new_unchecked function which assumes that the application has properly set up everything.

A `new` function was my initial approach, but was abandoned due to the potential combinatorial explosion in future support for all device families and FMC configurations (i.e., `device family <-> gpio ports mappings/AFs <-> possible FMC configurations w/r/t bus widths and other pins`). OTOH, I can totally see a separate `FMCAddressBus`, `FMCDataBus`, etc. set of traits which /could/ encapsulate the different non-swappable pins, but again, their construction would involve significant boilerplate and doesn't seem to me to bring much value unless the rest of FMC's more sophisticated capabilities are supported (e.g., NOR/PSRAM mux, etc.)

Reply:
> Agree with the risk, I was hoping it's mitigated by the fact that the operations they do are idempotent (i.e., they never disable any GPIO peripherals or pins, only additively enable them), and banking on the fact that the RegisterBlocks are guarded by VolatileCells in the upstream API

VolatileCell is not providing any protection against racy RMW operations.

> Do you mean the GPIOs which are passed into configure, or generally the notion of impl SDRAMConfig being ultimately responsible for setting up the GPIOs.

Both. The pins used by FMC/FSMC are the same, independent of the mode or devices connected to the bus. Probably we should even separate the bus setup altogether from the configuration of the channels.

> A new function was my initial approach, but was abandoned due to the potential combinatorial explosion in future support for all device families and FMC configurations (i.e., device family <-> gpio ports mappings/AFs <-> possible FMC configurations w/r/t bus widths and other pins).

It's not too bad, really. We'll have to deal with that anyway at some point.

I can test e.g. with a STM32F413H Disco (FSMC) and would certainly like to see the display work with this driver. I do have it running (in a hackish way) and can certainly provide some details about that.

Reply:
> VolatileCell is not providing any protection against racy RMW operations.

Someone would be using this code with ownership of the peripheral register blocks anyway when doing their GPIO setup, and the VolatileCell guards against races due to the compiler reordering the accesses. I suspect we might not be on the same page with what "racy" means, could you elaborate?

> Both. The pins used by FMC/FSMC are the same, independent of the mode or devices connected to the bus. Probably we should even separate the bus setup altogether from the configuration of the channels.

I just started doing this, but this does mean a separate type for each possible width of address bus (+3 types for data buses). To enforce stuff like valid values of Column vs Row bits, how do you feel about requiring `const_generics` to support this?

Reply:
> Someone would be using this code with ownership of the peripheral register blocks anyway when doing their GPIO setup

How do you mean? You're conjuring the PeripheralBlocks out of thin air with your `steal()`. Stealing != owning.

> I just started doing this, but this does mean a separate type for each possible width of address bus (+3 types for data buses).

Not necessarily. Depending on how we want to achieve that we could do a lot of different things, including automatically generating enums with all possible variants or specififying the maximum possible and allowing dummy pins.

> how do you feel about requiring const_generics to support that?

Anything not requiring nightly is fair game. 😅

Reply:
> How do you mean? You're conjuring the PeripheralBlocks out of thin air with your steal(). Stealing != owning.

You still need ownership of PeripheralBlocks to make the SCKEx, SDNEx, and SDNWE pins that you need in order to call `SDRAMConfig::configure`. So yes, on the surface I'm conjuring it up, but in practice, it's ultimately pointing the same instance of PeripheralBlocks that the caller owns anyway in order to be able to kick off those steals in the first place. Which is why I'm not sure how a race how would happen. 

> including automatically generating enums with all possible variants or specififying the maximum possible and allowing 

This sounds like we'd need proc macros, which I'm not against all, and might be the way forward as long as anything stable is fair game

Reply:
> You still need ownership of PeripheralBlocks to make the SCKEx, SDNEx, and SDNWE pins that you need in order to call SDRAMConfig::configure. So yes, on the surface I'm conjuring it up, but in practice, it's ultimately pointing the same instance of PeripheralBlocks that the caller owns anyway in order to be able to kick off those steals in the first place. Which is why I'm not sure how a race how would happen.

Not sure how you guarantee that everything is pointing to the same RegisterBlock, the pins could be very well on different ports. But that point is moot anyway because you're not proving anything at all by having access to a few pins because the registers are shared on purpose between multiple GPIOs which means that individual register access to stateful registers need to be guarded.

> This sounds like we'd need proc macros, which I'm not against at all, and might be the way forward as long as anything stable is fair game

You probably mean macros in general? Procedural macros would be a first in an HAL impl.  😅

Reply:
> Not sure how you guarantee that everything is pointing to the same RegisterBlock, the pins could be very well on different ports. But that point is moot anyway because you're not proving anything at all by having access to a few pins because the registers are shared on purpose between multiple GPIOs which means that individual register access to stateful registers need to be guarded.

I can prove it by reading the fine source code :P . I'm well aware it's a weak argument, I'm mostly just making a case for developer convenience at the cost of bending some absolute (general) guarantees which seem to hold anyway in this particular context.

> You probably mean macros in general?

I actually mean procedural macros. You'd need to do some subset of dependent types to get, for example, the number of row and column bits vs. address bus size, which you can approximate by implementing a subset of something like [Haskell's `singletons`](https://cs.brynmawr.edu/~rae/papers/2012/singletons/paper.pdf). (full disclosure, i'm one of *those* people :P). I suspect you'd need to end up needing to manipulate the AST in ways more sophisticated than `macro_rules` lets you do. Naturally, all that wouldn't affect anything at runtime, just there to let the typechecker keep you honest

> Procedural macros would be a first in an HAL impl

Sounds great for marketing :P

Reply:
> I can prove it by reading the fine source code :P . I'm well aware it's a weak argument, I'm mostly just making a case for developer convenience at the cost of bending some absolute (general) guarantees which seem to hold anyway in this particular context.

I don't see how you can claim there're no concurrent register accesses from reading the code. When creating such interfaces it is important to balance the "convenience" of not having to specify a boatload of specific pins against the possibility of incorrect use. I actually prefer the verbosity of the type system over the lack of any sanity checking (which is exactly what the C world is suffering quite a bit from).

It seems that other folks are strongly considering two versions, the "traditional" one with the verbosity and a `new_unchecked` which expects the user to setup everything correctly before going ahead and simply use it. I'd be open to such a split approach.

> I actually mean procedural macros.

I'm not sure we really need procedural macros but I'm also not opposed to using them. I'm just afraid it's going to be tedious due to the requirement to have the proc macros in a separate crate.

Reply:
> I don't see how you can claim there're no concurrent register accesses from reading the code.

I'm think I'm starting to see what you mean by races. I can think of only one instance where the non-atomicity of RMW would be an issue here -- an interrupt firing as we're in the middle of the setup process with someone's handler changing a pin on one of the GPIO peripherals we use to a High-Z state or something like that. Since this code is intended to be used as part of initialization, would it be acceptable to mitigate it by disabling interrupts at the start of `configure` and reenabling them before returning? We could also forgo stealing the FMC peripheral as it's wholly unnecessary to steal it -- and its only been stolen cause we had stolen access to the peripheral block anyway. 

> When creating such interfaces it is important to balance the "convenience" of not having to specify a boatload of specific pins against the possibility of incorrect use.

With you 100%. I just also want to note, much of this conversation for me is to see where the balance lies here, not just to defend my approach wholesale (though, admittedly, I may be just a tad biased :P)

Reply:
So, I just made a variant of this PR where you pass in all pins, (including address and data buses), and it seems to not actually fit in the 64K flash defined in `memory.x` for all the examples. Building with `--release` seems to make it fit, as does bumping the `FLASH = LENGTH 64K` to `128K` in `memory.x`, but this is a mere workaround for the bloat in compiled code that is introduced by this approach (i.e., forcing developers to use `--release` to use this may hamper their efforts in debugging, which seems unfavorable to me). I'm also not entirely sure where it originates, as not even `#[inline]`ing seems to resolve it. With all of that in mind, I'm starting to be convinced that stolen Address/Data buses and owned `FMC` and other support pins might be the only viable approach.

I defer to you to determine the best way forward here. Worth mentioning is after this is ironed out I'm planning to make a similar driver for the F429's `LTDC` peripheral which has a similar requirement with a wide GPIO bus necessary for its functionality (albeit, there are more pinswapping options, so it may require large numbers of traits, etc. which could create similar code bloat issues). All that to say, whatever is the approach taken here will likely end up informing that effort. Perhaps it makes more sense to divert this effort into @richardeoin's crate, so that building the examples isn't bounded by the presence of the `fmc` submodule here?

Reply:
`#[inline]` has no effect in dev mode. Need to go `#[inline(always)]` for that.

Reply:
Just peppered #[inline(always)] everywhere, same result.

Reply:
I've pushed an [`fmc-support-owned`](https://github.com/iostat/stm32f4xx-hal/tree/fmc-support-owned) branch if you want to see if you can make it fit. I'm going to make this PR rely less on unsafe in the interim w/r/t owning FMC and taking ownership of as many pins as possible.

Reply:
Need to go deeper into the rabbit hole and figure out where the bloat is coming from. It's unfortunate that dev mode is pretty much unusable for embedded but it's also not worth working around it in weird ways  because it's quite likely that you'll hit the next road block right around the corner (usually in the form of code executing so slow due a massive amount of silly function calls that all essential timing goals are missed).

Reply:
> I've pushed an [`fmc-support-owned`](https://github.com/iostat/stm32f4xx-hal/tree/fmc-support-owned) branch if you want to see if you can make it fit.

Making code work in dev mode is a tertiary goal for me at best. As I mentioned before I'd be okay with having two ways of initialising, one "regular" `new()` with all pins and one `new_unchecked()`, e.g. like done here: https://github.com/stm32-rs/stm32h7xx-hal/blob/master/src/spi.rs. Though I will insist on that being `unsafe` because I very much disagree on the `unchecked` postfix providing a clear enough message about the real unsafety (and that `unsafe` is only applicable to memory safety) of this.